### ISSUE TITLE
Stream recording on PulseAudio starts before startStream being called

### ISSUE THREAD

**Issue by dagargo** [thestk/rtaudio#401]
I've noticed that when using PulseAudio as the backend, an "internal" recording buffer is filled up before `startStream` is being called.

Audio playback seems to work as expected.
Other backends seem to record as expected too (tested on LINUX_PULSE and WINDOWS_WASAPI).
This happens at least since 5.2.0 and the behavior is present in the master branch too.

This minimal patch over the included test `record.cpp` provides the proof. Notice the 5 s sleep.

```
$ git diff record.cpp
diff --git a/tests/record.cpp b/tests/record.cpp
index c5f9cb5..2587247 100644
--- a/tests/record.cpp
+++ b/tests/record.cpp
@@ -111,7 +111,7 @@ int main( int argc, char *argv[] )
   // minimal command-line checking
   if ( argc < 3 || argc > 6 ) usage();
 
-  RtAudio adc;
+  RtAudio adc = RtAudio(RtAudio::LINUX_PULSE);
   std::vector<unsigned int> deviceIds = adc.getDeviceIds();
   if ( deviceIds.size() < 1 ) {
     std::cout << "\nNo audio devices found!\n";
@@ -165,6 +165,9 @@ int main( int argc, char *argv[] )
     goto cleanup;
   }
 
+  sleep (5);
+  std::cout << "Start recording...";
+
   if ( adc.startStream() ) goto cleanup;
 
   std::cout << "\nRecording for " << time << " seconds ... writing file 'record.raw' (buffer frames = " << bufferFrames << ")." << std::endl;
```

When asking the included script `record` for a 5 s sample, this is the result.

```
tests$ ./record 2 48000 5
Start recording...
Recording for 5 seconds ... writing file 'record.raw' (buffer frames = 512).
```

Obviously, there is a pause of 5 s before the first line is printed. Then, there is no pause between the first message and the second one and the execution ends. The callback is invoked at the moment of calling `startStream` as many times as needed to provide the client the with the 5 s of recorded audio since `openStream`.

Can anyone replicate this or am I missing something?

Before giving it a try, I need to be sure the error is not on my side.

**Comment by garyscavone** [comment]
Or change it to the setting for playback on line 9539.

**Comment by garyscavone** [comment]
I just tested this. I verified the behavior you described initially. But when I made the suggested change, I found that the record.cpp test program never finishes if maxlength is set to bufferBytes. I had to make maxlength at least equal to 2 * bufferBytes to work. I am committing a change that fixes this but I will use (options->numberOfBuffers+1) if options are provided and bufferBytes * 4 otherwise. I just pushed the change to the repo.
